python/cuml/experimental/fil/fil.pyx

```diff
@@ -1228,6 +1263,106 @@ class ForestInference(UniversalBase, CMajorInputTagMixin):
+        it is not, random data will be generated based on the indicated batch
+        size. After finding the optimal layout, the model will be reloaded if
+        necessary. The optimal chunk size will be used to set the default chunk
+        size used if none is passed to the predict call.
+
+        Parameters
+        ----------
+        data
+            Example data of shape iterations x batch size x features or None.
+            If None, random data will be generated instead.
+        batch_size : int
+            If example data is not provided, random data with this many rows
+            per batch will be used.
+        iterations : int
+            The number of times to repeat inference in order to mitigate
+            outliers in performance data.
+        predict_method : str
+            If desired, optimization can occur over one of the prediction
+            method variants (e.g. "predict_per_tree") rather than the
+            default `predict` method. To do so, pass the name of the method
+            here.
+        max_chunk_size : int or None
+            The maximum chunk size to explore during optimization. If not
+            set, a value will be picked based on the current device type.
+            Setting this to a lower value will reduce the optimization search
+            time but may not result in optimal performance.
+        seed : int
+            The random seed used for generating example data if none is
+            provided.
+        """
+        if data is None:
+            xpy = GlobalSettings().xpy
+            dtype = self.forest.get_dtype()
+            data = xpy.random.uniform(
+                xpy.finfo(dtype).min / 2,
+                xpy.finfo(dtype).max / 2,
+                (iterations, batch_size, self.forest.num_features())
+            )
+        else:
+            data = CumlArray.from_input(
+                data,
+                order='K',
+            ).to_output('array')
+        try:
+            iterations, batch_size, features = data.shape
+        except ValueError:
+            batch_size, features = data.shape
+            data = [data for _ in range(iterations)]
+
+        if max_chunk_size is None:
+            max_chunk_size = 512
+        if GlobalSettings().device_type == DeviceType.device:
+            max_chunk_size = min(max_chunk_size, 32)
+
+        infer = getattr(self, predict_method)
+
+        optimal_layout = 'depth_first'
+        optimal_chunk_size = 1
+        optimal_time = float('inf')
+
+        for layout in ('depth_first', 'breadth_first'):
+            self.layout = layout
+            chunk_size = 1
+            while chunk_size <= max_chunk_size:
+                # Warmup
+                infer(data[0], chunk_size=chunk_size)
+                # Measure performance
+                start = perf_counter()
+                for iter_index in range(iterations):
+                    infer(
+                        data[iter_index], chunk_size=chunk_size
```

Comment:
Do we have to use the index or could we just infer the same data each time?

Reply:
My thought is that it's better to have a bit of variety in the data, particularly if someone is optimizing for small batch size. Particularly with the depth-first layout or with shallow trees, it's possible to get lucky in your data and blaze down the "happy path" as you're traversing the tree. In that case, you may find an optimum that is specific to those data but which may not generalize. For small models on few features, there's also the possibility that you'll end up with the entirety of the input data that you need in cache, which would also change the results.

I doubt that it makes much difference in the vast majority of cases, but I'm trying to err on the side of caution with a method that is intended to be more or less blindly applied. Seem reasonable or should I think about this one more?

Reply:
That's fair, but we are literally just copying the same data in line 1334, are we not?